pkg/nats/proxy/callback_handler.go

```diff
@@ -0,0 +1,82 @@
+package proxy
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+	"strings"
+
+	"github.com/bacalhau-project/bacalhau/pkg/compute"
+	"github.com/nats-io/nats.go"
+	"github.com/rs/zerolog/log"
+)
+
+type CallbackHandlerParams struct {
+	Name     string
+	Conn     *nats.Conn
+	Callback compute.Callback
+}
+
+// CallbackHandler is a handler for callback events that registers for incoming nats requests to Bacalhau callback
+// protocol, and delegates the handling of the request to the provided callback.
+type CallbackHandler struct {
+	name     string
+	conn     *nats.Conn
+	callback compute.Callback
+}
+
+type callbackHandler[Request any] func(context.Context, Request)
+
+func NewCallbackHandler(params CallbackHandlerParams) (*CallbackHandler, error) {
+	handler := &CallbackHandler{
+		name:     params.Name,
+		conn:     params.Conn,
+		callback: params.Callback,
+	}
+
+	subject := callbackSubscribeSubject(handler.name)
+	_, err := handler.conn.Subscribe(subject, func(m *nats.Msg) {
+		handleCallback(m, handler)
+	})
+	if err != nil {
+		return nil, err
+	}
+	log.Debug().Msgf("ComputeHandler %s subscribed to %s", handler.name, subject)
+	return handler, nil
+}
+
+// handleRequest handles incoming NATS messages.
+func handleCallback(msg *nats.Msg, handler *CallbackHandler) {
```

Comment:
Should this be a method on *CallbackHandler?

Reply:
yeah that works since this method doesn't use generics like `compute_handler`. Fixed in https://github.com/bacalhau-project/bacalhau/pull/3188/commits/a15b80d84c7cf2e9354bc939b3219468346014c5